src/Microsoft.AspNet.OData/Query/TextPlainODataQueryOptionsParser.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using Microsoft.OData;
+using Microsoft.AspNet.OData.Common;
+
+namespace Microsoft.AspNet.OData.Query
+{
+    public partial class TextPlainODataQueryOptionsParser
+    {
+        /// <inheritdoc/>
+        public string Parse(Stream requestStream)
+        {
+            string queryString = string.Empty;
+
+            MemoryStream memoryStream = new MemoryStream();
+            TextReader reader;
+
```

Comment:
Just to preempt possible questions around this seemingly "unattractive" approach, the following "attractive" approaches are flagged by the analyzer in **Microsoft.AspNet.OData** project. The analyzer in **Microsoft.AspNetCore.OData** however didn't appear to make a fuss.

1. [CA2202: Do not dispose object multiple times](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2202-do-not-dispose-objects-multiple-times?view=vs-2015) against the following two approaches:

    a. Approach # 1
    ```c#
    using (var memoryStream = new MemoryStream())
    {
        requestStream.CopyTo(memoryStream);
        memoryStream.Position = 0;
    
        using (var reader = new StreamReader(memoryStream))
        {
            string result = reader.ReadToEnd();
        }
    }
    ```

    b. Approach # 2
    ```c#
    var memoryStream = new MemoryStream();

    try
    {
        requestStream.CopyTo(memoryStream);
        memoryStream.Position = 0;
    
        using (var reader = new StreamReader(memoryStream))
        {
            string result = reader.ReadToEnd();
        }
    }
    catch
    {
        // CopyTo() may throw ObjectDisposedException
        memoryStream.Dispose();
    }
    ```

2. [CA2000: Dispose objects before losing scope](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2000-dispose-objects-before-losing-scope?view=vs-2015) against the following approach:
    ```c#
    var memoryStream = new MemoryStream();

    using (var reader = new StreamReader(memoryStream))
    {
        requestStream.CopyTo(memoryStream);
        memoryStream.Position = 0;
    
        string result = reader.ReadToEnd();
    }
    ```

Reply:
What about:
```c#
var memoryStream = new MemoryStream();
try {
    using(var reader = new StreamReader(memoryStream))
    {
    // do your parsing ...
    }
}
catch
{
    throw new ODataException(SRResources.CannotParseQueryOptionsPayload);
}
finally
{
    if (memoryStream != null) memoryStream.Dispose();
}

Reply:
@gathogojr, it is possible to suppress analyzers when you have a valid justification...

Reply:
@g2mula In this case I'm not 100% sure that I have a valid justification to suppress the warning so it might be better to err on the side of caution. I have used the nested `using`s liberally before but as I scrutinized it this time, I realized the argument by the analyzer would be valid for certain scenarios.

I think the reason why we're able to get away with it is best captured in [this answer](https://stackoverflow.com/a/4274769/752739) to a StackOverflow post I came across - `MemoryStream` does not have any unmanaged resources to dispose. Plus unlike other streams, I'd expect `MemoryStream` to be less prone to typical failures